src/main/java/uk/co/westhawk/snmp/stack/SnmpContextv3Basis.java

```diff
@@ -1179,15 +1260,87 @@ public String toString()
     buffer.append(", userAuthenticationPassword=").append(userAuthenticationPassword);
     buffer.append(", usePrivacy=").append(usePrivacy);
-    buffer.append(", privacyProtocol=").append(ProtocolNames[privacyProtocol]);
+    buffer.append(", privacyProtocol=").append(PROTOCOL_NAMES[privacyProtocol]);
     buffer.append(", userPrivacyPassword=").append(userPrivacyPassword);
     buffer.append(", #trapListeners=").append(trapSupport.getListenerCount());
     buffer.append(", #pduListeners=").append(pduSupport.getListenerCount());
     buffer.append("]");
     return buffer.toString();
 }
 
+	/**
+	 * Generates the privacy key based on the authentication protocol.
+	 *
+	 * @param context The SNMP context.
+	 * @param engineId The SNMP engine ID.
+	 * @param authenticationProtocol The authentication protocol.
+	 * @return The generated privacy key.
+	 */
+	protected byte[] generatePrivacyKey(SnmpContextv3Basis context, String engineId, int authenticationProtocol) {
+		byte[] derivedPrivacyKey;
+		byte[] localizedPrivacyKey = null;
+		if (authenticationProtocol == context.MD5_PROTOCOL) {
+			derivedPrivacyKey = context.getPrivacyPasswordKeyMD5();
+			localizedPrivacyKey = SnmpUtilities.getLocalizedKeyMD5(derivedPrivacyKey, engineId);
+		} else if (authenticationProtocol == context.SHA1_PROTOCOL) {
+			derivedPrivacyKey = context.getPrivacyPasswordKeySHA1();
+			localizedPrivacyKey = SnmpUtilities.getLocalizedKeySHA1(derivedPrivacyKey, engineId);
+		} else if (authenticationProtocol == context.SHA256_PROTOCOL) {
+			derivedPrivacyKey = context.getPrivacyPasswordKeySHA256();
+			localizedPrivacyKey = SnmpUtilities.getLocalizedKeySHA256(derivedPrivacyKey, engineId);
+		} else if (authenticationProtocol == context.SHA512_PROTOCOL) {
+			derivedPrivacyKey = context.getPrivacyPasswordKeySHA512();
+			localizedPrivacyKey = SnmpUtilities.getLocalizedKeySHA512(derivedPrivacyKey, engineId);
+		} else if (authenticationProtocol == context.SHA224_PROTOCOL) {
+			derivedPrivacyKey = context.getPrivacyPasswordKeySHA224();
+			localizedPrivacyKey = SnmpUtilities.getLocalizedKeySHA224(derivedPrivacyKey, engineId);
+		} else if(authenticationProtocol == context.SHA384_PROTOCOL) {
+			derivedPrivacyKey = context.getPrivacyPasswordKeySHA384();
+			localizedPrivacyKey = SnmpUtilities.getLocalizedKeySHA384(derivedPrivacyKey, engineId);
+		}
+		return localizedPrivacyKey;
+	}
+
+	/**
+	 * Computes the fingerprint for the given SNMP message.
+	 *
+	 * @param context The SNMP context.
+	 * @param snmpEngineId The SNMP engine ID.
+	 * @param authenticationProtocol The authentication protocol.
+	 * @param computedFingerprint The computed fingerprint.
+	 * @param message The SNMP message.
+	 * @return The computed fingerprint.
+	 */
+	protected byte[] computeFingerprint(SnmpContextv3Basis context, String snmpEngineId, int authenticationProtocol, byte[] computedFingerprint, byte[] message) {
```

Comment:
Remove `SnmpContextv3Basis context` from the method arguments. We are already in the current context.